WireDomain/Sources/WireDomain/Event Processing/ConversationEventProcessor/ConversationAccessUpdateEventProcessor.swift

```diff
@@ -26,15 +27,48 @@ protocol ConversationAccessUpdateEventProcessorProtocol {
     ///
     /// - Parameter event: A conversation access update event.
 
-    func processEvent(_ event: ConversationAccessUpdateEvent) async throws
+    func processEvent(_ event: ConversationAccessUpdateEvent) async
 
 }
 
 struct ConversationAccessUpdateEventProcessor: ConversationAccessUpdateEventProcessorProtocol {
 
-    func processEvent(_: ConversationAccessUpdateEvent) async throws {
-        // TODO: [WPB-10164]
-        assertionFailure("not implemented yet")
+    let context: NSManagedObjectContext
+    let repository: any ConversationRepositoryProtocol
+
+    func processEvent(_ event: ConversationAccessUpdateEvent) async {
+        let conversationID = event.conversationID
+
+        let localConversation = await repository.fetchOrCreateConversation(
+            with: conversationID.uuid,
+            domain: conversationID.domain
+        )
+
+        let accessRoles = if let legacyAccessRole = event.legacyAccessRole {
+            getAccessRoles(from: legacyAccessRole)
+        } else {
+            event.accessRoles
+        }
+
+        await context.perform {
+            localConversation.accessModeStrings = event.accessModes.map(\.rawValue)
+            localConversation.accessRoleStringsV2 = accessRoles.map(\.rawValue)
```

Comment:
context.save() ?

Reply:
I think saves should happen on a higher level. For instance:

- Get batch of pending events
- Process batch of events (no saves in-between)
- After successfully processing all events in batch, one single save, then delete those pending events (they are consumed now

We need to take care that any committed work isn't repeated in case of any interruption to the event processing.

Reply:
Just sharing my thoughts regarding these topics here:

**1-** Saving on a higher level would imply to delegate the responsibility of persisting CoreData changes to another layer/object (whatever that layer/object is) whereas IMO it should be the sole responsibility of the local stores (which will encapsulate all of our storage operations and should be responsible of actually saving these changes to the persistent store when needed _or when asked for it_). 

--> what we want to avoid is having context saving spread everywhere in codebase (we might end up with funny behaviors if somewhere we forget to save the changes). that's more error prone.

--> we want to be our different layers to be well-isolated and decoupled, for instance I don't think we want our `SyncManager` to access a CoreData context directly but go through an intermediate object for instance a `LocalStoreManager` that owns multiple local stores and can be responsible of handling global "decisions" on the persistent store (e.g `didFinishSlowSync()` `didFinishQuickSync()` ) and saving the context or rolling back latest commits.

**2-** Do we want to wait for all events to be processed to save the changes (I understand we might want to to so for performance reason) but that means if we have 8 events to process and 7 are successfully processed but the last one is not, we will discard all committed changes and redo the work all over again for 8 events even if only one failed. 

Also, even if we consider events we receive as a batch, they don't seem interdependent to me. So we could consider saving each one of them after they're being processed, so if we have to process a batch of 4 events, we'll process 1. then 2. then 3. then 4. and if the 2. fails we could still process 3. and 4. since they don't depend on 2.). Then, saving the context after each event is processed makes sense (I think).

Also, we want the UI to  _always_ reflect the latest database "state" so it could be risky to discard a batch of 100 events all at once.

It's true though that it comes at a cost, performance wise this is not ideal.. especially if we have a lot of events to process.